### PR TITLE
Upgrade TargetPlatformVersion to 10.0.19041.0

### PR DESCRIPTION
**Pull request by JunielKatarn**
Version `10.0.19041.0` is the current default for Visual Studio.
As far as I know, current targeted version `10.0.18362.0` does not have any benefit over `10.0.19041.0`, and takes an additional couple gigabytes because the latter is installed mandatorily.

This should also speed up dev and CI dependency installation.

Requesting backport because otherwise we would still require `10.0.18362.0` to build stable branches.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/7779)

### CODE REVIEW

**Review by NickGerleman — APPROVED** [review]
Looks good, but we may not want to backport something that changes local requirements.

File: packages/playground/windows/playground-win32/Playground-win32.vcxproj
Comment (asklar) [inline packages/playground/windows/playground-win32/Playground-win32.vcxproj:10]: Please don't change the min version

File: vnext/Microsoft.ReactNative.Cxx/ReactPropertyBag.h
Comment (asklar) [inline vnext/Microsoft.ReactNative.Cxx/ReactPropertyBag.h:46]: Replace with version and language agnostic url:
```suggestion
// https://docs.microsoft.com/uwp/api/windows.foundation.propertytype
```

File: vnext/Microsoft.ReactNative.Managed.CodeGen.UnitTests/AnalysisTestBase.cs
Comment (asklar) [inline vnext/Microsoft.ReactNative.Managed.CodeGen.UnitTests/AnalysisTestBase.cs:93]: I don't follow some of the assertions in the comment but where is this #define coming from? can't find it defined anywhere and now it would be mismatched with the value below

File: vnext/PropertySheets/React.Cpp.props
Comment (asklar) [inline vnext/PropertySheets/React.Cpp.props:7]: this is no longer 19H1

File: vnext/PropertySheets/React.Cpp.props
Comment (asklar) [inline vnext/PropertySheets/React.Cpp.props:18]: why were these removed?

**Review by asklar — CHANGES_REQUESTED** [review]
This could be somewhat of a breaking change for people who have already set up their environment; if they have 18362 sdk installed and now we start requiring 19041. We should be judicious with new requirements like this one, so I don't think we should backport it, and we should mark this as a breaking change.

Left some comments on things I believe are bugs from the find/replace.

File: packages/playground/windows/playground-win32/Playground-win32.vcxproj
Comment (JunielKatarn) [reply packages/playground/windows/playground-win32/Playground-win32.vcxproj:10]: Not intended.

Fixing...

File: vnext/PropertySheets/React.Cpp.props
Comment (JunielKatarn) [reply vnext/PropertySheets/React.Cpp.props:18]: Not removed, just placed in the correct PROPS file.

File: vnext/Microsoft.ReactNative.Managed.CodeGen.UnitTests/AnalysisTestBase.cs
Comment (JunielKatarn) [reply vnext/Microsoft.ReactNative.Managed.CodeGen.UnitTests/AnalysisTestBase.cs:93]: Neither did I.

May be an outdated and possibly no longer used macro.

I'll ping the core team about this before merging.

**Comment by JunielKatarn** [comment]
> This could be somewhat of a breaking change for people who have already set up their environment; if they have 18362 sdk installed and now we start requiring 19041. We should be judicious with new requirements like this one, so I don't think we should backport it, and we should mark this as a breaking change.
> 
> Left some comments on things I believe are bugs from the find/replace.

As of VS 16.9, the SDK `19041` is installed by default with the UWP development toolkit. No supported way around that that I know of.
Dropping the previous version allows anyone working on master and soon 0.65 to not install the older Win SDK.

Should we assume the majority of RNW consumers stay with older versions of VS (i.e. 16.7) that mandatorily install an older SDK?

**Comment by asklar** [comment]
main concern is folks who have been using RNW and have the 18362 SDK installed; now this update will require them to install the 19041 SDK, how do we help them before they are broken?

**Comment by JunielKatarn** [comment]
> main concern is folks who have been using RNW and have the 18362 SDK installed; now this update will require them to install the 19041 SDK, how do we help them before they are broken?

Just to be clear, that would only happen if they were stalling the VS update and still use an older VS version (i.e. 16.7).

Anyone using 16.9 will get the 19041 SDK installed by default when adding UWP functionality so the change would be transparent.

It boils down to what is the most likely version installed by RNW consumers, in my opinion.

**Review by asklar — APPROVED** [review]

**Comment by brodybits** [comment]
> Those who don't have it installed is only because they are not on VS 16.9. Wouldn't this be a reasonable requirement?

Considering the following note in <https://microsoft.github.io/react-native-windows/docs/rnw-dependencies>:

> Install [Visual Studio 2019 (version 16.5 or greater)](https://www.visualstudio.com/downloads)

I would consider suddenly requiring some VS pre-16.9 versions to install or upgrade something to be a breaking change.

**Comment by brodybits** [comment]
Maybe I am getting excessively pedantic here, considering the number of platform tools with breaking changes in non-0.x minor release numbers. I hope we can update [System Requirements / RNW dependencies (1)](https://microsoft.github.io/react-native-windows/docs/rnw-dependencies) to better reflect the potential divergence between RNW and VS 2019, like I tried proposing in [microsoft/react-native-windows-samples#443 (2)](https://github.com/microsoft/react-native-windows-samples/pull/443).

[1] https://microsoft.github.io/react-native-windows/docs/rnw-dependencies
[2] https://github.com/microsoft/react-native-windows-samples/pull/443

**Comment by asklar** [comment]
As we evolve the product and bugs in the compiler, tools, or underlying platform are fixed, we will need to update what VS version is the minimum required. VS is fairly aggressive with auto-updating, ADO/GH CI updates the version available to run within a week or two, so we stop building with older compiler/linkers pretty soon after a VS release is marked as public. In reality we can only ever support the latest VS. I agree that the SDK update is a breaking change we should call out if this gets into 0.65 (@NickGerleman  fyi as he is collating these).

I think it's fine to not use the "by-default installed SDK with the VS release", as long as we call out _which_ SDK to install. 19041 has some MIDL improvements we'd benefit from. Today the RNW docs should be calling out the 18362 SDK, after this change we should call out 19041, so maybe that should be the right change to make for microsoft/react-native-windows-samples#443.

As far as the concern about breaking changes, I'm not sure if you meant breaking changes in VS tooling/compilers, or in RNW? If  you're referring to RNW, we follow the RN core versioning so it's still a 0.x release.

**Comment by brodybits** [comment]
> In reality we can only ever support the latest VS.

Makes sense to me (should probably be part of the documentation update).

> I agree that the SDK update is a breaking change we should call out if this gets into 0.65

It looks to me like react-native-windows minor version number had been kept (roughly ?) in sync with the core react-native minor version number, and I hope this would not be broken. I start to wonder if we could have a cut of both react-native 0.65 and react-native-windows 0.65 in the near future, perhaps starting with (pre-)beta?

> I think it's fine to not use the "by-default installed SDK with the VS release", as long as we call out _which_ SDK to install.

This would be nice, (but) would require more (manual ?) work to keep the code and documentation in sync. The general note that I proposed in <https://github.com/microsoft/react-native-windows-samples/pull/443> was meant to help avoid this kind of extra work and potential issues that could result in the future.

> As far as the concern about breaking changes, I'm not sure if you meant breaking changes in VS tooling/compilers, or in RNW?

By "number of platform tools with breaking changes in non-0.x minor release numbers" I meant breaking changes in the VS tooling & compilers, as well as breaking changes in tooling for Android such as Android Studio and Gradle.

**Comment by asklar** [comment]
Yes, we are aligned to the core RN versioning and try to release very shortly after RN officially releases (0.64 RNW shipped the day after RN!). FYI that we created the 0.65 branch yesterday; changes that we want to put into 0.65 will now need to be backported so by default this change will only go into 0.66+ (@JunielKatarn fyi).

**Comment by jonthysell** [comment]
I don't think it makes sense to backport a bump in SDK versions to older releases. We have to accept that VS only offers downloads for it's latest IDE, but the SDK will always still be available.

Imagine you're an existing dev on a current 63/64 app. Any future backports (which normally is only super important non-breaking things like security updates) will require them to retarget their app just to get future updates.

If you're a new dev and for some reason want to target not the latest version of RNW with a "new" app, then I think having a call out that you'll need to install an older SDK to work on an older version isn't that crazy, and doesn't break anything.

**Comment by JunielKatarn** [comment]
> I don't think it makes sense to backport a bump in SDK versions to older releases. We have to accept that VS only offers downloads for it's latest IDE, but the SDK will always still be available.
> 
> Imagine you're an existing dev on a current 63/64 app. Any future backports (which normally is only super important non-breaking things like security updates) will require them to retarget their app just to get future updates.
> 
> If you're a new dev and for some reason want to target not the latest version of RNW with a "new" app, then I think having a call out that you'll need to install an older SDK to work on an older version isn't that crazy, and doesn't break anything.

Point taken.
That said, modern VS/MSBuild allows to target the "latest available" SDK by using a shorthand value of `10.0` instead of the full SDK version.

Maybe we can rely on this in the future, but I'll explore that approach after merging this change.

**Comment by JunielKatarn** [comment]
@dannyvv @NickGerleman Would you happen to know what this error is about?
Else, do you know how to reproduce this (build `packages\integration-test-app\windows\integrationtest.sln`) locally?

See `error C3246` at https://dev.azure.com/ms/react-native-windows/_build/results?buildId=169417&view=logs&j=ce6dbd70-21a8-576a-2d1e-b860511db24a&t=99422dc1-32ec-55ed-17e7-9f92b00603bf&l=7406

**Comment by asklar** [comment]
> See `error C3246` at https://dev.azure.com/ms/react-native-windows/_build/results?buildId=169417&view=logs&j=ce6dbd70-21a8-576a-2d1e-b860511db24a&t=99422dc1-32ec-55ed-17e7-9f92b00603bf&l=7406
> 
> @dannyvv @NickGerleman Would you happen to know what this error is about?
> Else, do you know how to reproduce this (build `packages\integration-test-app\windows\integrationtest.sln`) locally?

Likely this is due to a new VS ingestion and unrelated to this change (but still needs to be fixed), can you upgrade your VS and try building locally to repro? it looks like some class is marked final when it shouldn't be.

**Comment by brodybits** [comment]
Any pointers how I could test this in my own React Native application?

I tried this command to add RNW from GitHub:

```console
npx react-native-windows-init --overwrite --version github:jurocha-ms/react-native-windows#winsdk-19041
```

and it ended with an error. Here is my output from the shell:

```console
PS C:\Users\brodybits\scoop\apps\uwptest200> npx react-native-windows-init --overwrite --version github:jurocha-ms/react-native-windows#winsdk-19041
npx: installed 129 in 7.734s
Reading project name from package.json...
Warning: Querying npm to find react-native-windows@github:jurocha-ms/react-native-windows#winsdk-19041 failed.  Attempting to continue anyway...
Installing react-native-windows@github:jurocha-ms/react-native-windows#winsdk-19041...
warning "react-native > react-native-codegen > jscodeshift@0.11.0" has unmet peer dependency "@babel/preset-env@^7.1.6".warning "@react-native-community/eslint-config > @typescript-eslint/eslint-plugin > tsutils@3.21.0" has unmet peer dependency "typescript@>=2.8.0 || >= 3.2.0-dev || >= 3.3.0-dev || >= 3.4.0-dev || >= 3.5.0-dev || >= 3.6.0-dev || >= 3.6.0-beta || >= 3.7.0-dev || >= 3.7.0-beta".
react-native-windows@0.0.1 successfully installed.
Cannot find module 'react-native-windows/local-cli/generate-windows'
Require stack:
- C:\Users\brodybits\AppData\Roaming\npm-cache\_npx\3804\node_modules\react-native-windows-init\lib-commonjs\requireGenerateWindows.js
- C:\Users\brodybits\AppData\Roaming\npm-cache\_npx\3804\node_modules\react-native-windows-init\lib-commonjs\Cli.js
- C:\Users\brodybits\AppData\Roaming\npm-cache\_npx\3804\node_modules\react-native-windows-init\bin.js
Error: Cannot find module 'react-native-windows/local-cli/generate-windows'
Require stack:
- C:\Users\brodybits\AppData\Roaming\npm-cache\_npx\3804\node_modules\react-native-windows-init\lib-commonjs\requireGenerateWindows.js
- C:\Users\brodybits\AppData\Roaming\npm-cache\_npx\3804\node_modules\react-native-windows-init\lib-commonjs\Cli.js
- C:\Users\brodybits\AppData\Roaming\npm-cache\_npx\3804\node_modules\react-native-windows-init\bin.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:885:15)
    at Function.resolve (internal/modules/cjs/helpers.js:94:19)
    at Object.requireGenerateWindows [as default] (C:\Users\brodybits\AppData\Roaming\npm-cache\_npx\3804\node_modules\react-native-windows-init\src\requireGenerateWindows.ts:24:33)
    at C:\Users\brodybits\AppData\Roaming\npm-cache\_npx\3804\node_modules\react-native-windows-init\src\Cli.ts:529:51
    at processTicksAndRejections (internal/process/task_queues.js:95:5) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    'C:\\Users\\brodybits\\AppData\\Roaming\\npm-cache\\_npx\\3804\\node_modules\\react-native-windows-init\\lib-commonjs\\requireGenerateWindows.js',
    'C:\\Users\\brodybits\\AppData\\Roaming\\npm-cache\\_npx\\3804\\node_modules\\react-native-windows-init\\lib-commonjs\\Cli.js',
    'C:\\Users\\brodybits\\AppData\\Roaming\\npm-cache\\_npx\\3804\\node_modules\\react-native-windows-init\\bin.js'
  ]
}
```

**Comment by asklar** [comment]
I don't believe we support the `github:` syntax. We have instructions for how to test changes to the CLI locally (you have to have a local clone of RNW) here https://github.com/microsoft/react-native-windows/wiki/Contributing-to-the-CLI

**Comment by asklar** [comment]
```
D:\a\1\s\packages\integration-test-app\windows\integrationtest\Generated Files\winrt\base.h(7717,29): error C2338: C++/WinRT implementation types must not be final (compiling source file TestHostHarness.cpp) [D:\a\1\s\packages\integration-test-app\windows\integrationtest\integrationtest.vcxproj]
D:\a\1\s\packages\integration-test-app\windows\integrationtest\TestHostHarness.cpp(140): message : see reference to function template instantiation 'winrt::com_ptr<D> winrt::make_self<IntegrationTest::TestTransaction,>(void)' being compiled [D:\a\1\s\packages\integration-test-app\windows\integrationtest\integrationtest.vcxproj]
```

The error message is coming from cppwinrt, when `WINRT_NO_MAKE_DETECTION` is not defined. I don't see anything in this PR related to that macro directly, but I wonder if the new SDK has it set somewhere. Can you produce a binlog of that project so we can take a look at who is setting it? @JunielKatarn

**Comment by JunielKatarn** [comment]
> ```
> D:\a\1\s\packages\integration-test-app\windows\integrationtest\Generated Files\winrt\base.h(7717,29): error C2338: C++/WinRT implementation types must not be final (compiling source file TestHostHarness.cpp) [D:\a\1\s\packages\integration-test-app\windows\integrationtest\integrationtest.vcxproj]
> D:\a\1\s\packages\integration-test-app\windows\integrationtest\TestHostHarness.cpp(140): message : see reference to function template instantiation 'winrt::com_ptr<D> winrt::make_self<IntegrationTest::TestTransaction,>(void)' being compiled [D:\a\1\s\packages\integration-test-app\windows\integrationtest\integrationtest.vcxproj]
> ```
> 
> The error message is coming from cppwinrt, when `WINRT_NO_MAKE_DETECTION` is not defined. I don't see anything in this PR related to that macro directly, but I wonder if the new SDK has it set somewhere. Can you produce a binlog of that project so we can take a look at who is setting it? @JunielKatarn

Only noticed this comment. I had removed the `final` qualifier from class `TestTransaction`, which solved the issue locally.
@NickGerleman  is this an acceptable solution, or is the qualifier crucial to the test infrastructure? See commit 49ad562b0f2c9fb0e00853bd6db22fc88f8c00fd.
Note, I'd still make sure the code builds as is with the old SDK.

**Comment by asklar** [comment]
yes removing the final keyword would get rid of the error, but it would be important to understand why we get the error with the new sdk, since the version of cppwinrt that we use is not the one that comes in the SDK, unless it is for this test project?

**Comment by JunielKatarn** [comment]
> yes removing the final keyword would get rid of the error, but it would be important to understand why we get the error with the new sdk, since the version of cppwinrt that we use is not the one that comes in the SDK, unless it is for this test project?

This IS a test project, so I'll go ahead and merge today.
That said, I agree that we should understand this behavior and will still generate detailed logs for building both with the "old" SDK and the default one.

**Comment by NickGerleman** [comment]
Need to update https://github.com/microsoft/react-native-windows-samples/edit/main/docs/win10-compat.md once we cut 0.65 versioned docs, since this change was in 0.66.

**Pull request merged**
